Album info

TotalDarkness-NRF opened the pull request:
This pull request focuses on adding a button and ui to albums to display some album info.

- Added a button in albums to go to the info screen
- Added a ui to display the info for albums

The layout is as follows:
![image](https://user-images.githubusercontent.com/63280150/129470963-a67aaea7-0ac5-40fc-83f2-dbc714a88657.png)

xou816 commented:
Hi! Thanks for being so invested in the project!! :) 

For large UI changes or additions though, I'd rather prefer an issue to begin with so that there's room to discuss what to do before you get too much work done!

I think it's a good idea to have a separate screen if we have lots of info to show, although maybe a modal would be better in that case :) 

Speaking of infos to display, I feel like some of them are either redundant (album/artist name) or not very relevant to display (markets). Also, I feel like there's room for extra infos on the existing album details screen -- maybe the release date, for instance, or the duration. 

I'd keep the label and copyright for starters; release date and duration too, unless we can fit them in the main screen; and album/artist name could be useful, but if we go with a modal I don't think re-contextualization is needed.

Don't apply those changes yet, let's discuss first! :) What do you think?

TotalDarkness-NRF commented:
Thanks for the feedback. I feel like the info would feel incomplete without showing artist and album name. As for markets maybe instead of removing it, hide it under something so if someone would like to see if the album is available for them they can.
I am not to sure what you mean by using a modal instead. I do use a model similar to how other screen are created and managed.
As for putting more info on the album screen, I would like to do that for a few items, but I am not sure how to do so without making it feel cluttered or redundant.

xou816 commented:
I meant a dialog, sorry! Rough Glade "sketch":
![Capture d’écran de 2021-08-09 19-12-23](https://user-images.githubusercontent.com/6277494/128746414-bc71c5eb-f4c4-459b-b738-0a17eca3a97b.png)

Since a dialog overlaps the current view, it is contextual and we don't need to repeat every bit of information -- although I think I agree with you that it does not hurt to have a reminder of the album/artist here! :) 

But I really miss the point of having ID, "is liked", markets, etc on that screen

TotalDarkness-NRF commented:
Thanks for the sketch. I will look at implementing this. This would be simpler then how I implemented it. As for markets, id, and liked, they wont be needed if I implement it like this. Those were just extra info I would have included just for a bit of extra "stuff"

TotalDarkness-NRF commented:
This is sort of what I have come up with. It contains the info in a dialogue modal as a pop up. Let me know what you think 
![Screenshot_20210809_192127](https://user-images.githubusercontent.com/63280150/128786587-354ad2c2-a1e4-48b3-882f-0f76fc99208b.png)

jacksongoode commented:
@TotalDarkness-NRF, could you post a screenshot of the default theme with that new modal?

jacksongoode commented:
> > @TotalDarkness-NRF, could you post a screenshot of the default theme with that new modal?
> > @jacksongoode Not sure what the default theme is but here is for the adwaita theme for light and dark
> 
> ![Screenshot_20210811_023754](https://user-images.githubusercontent.com/63280150/128981570-7a642afe-76cb-4313-82bd-d680db590fe4.png)
> ![Screenshot_20210811_023728](https://user-images.githubusercontent.com/63280150/128981567-1918308f-b8dc-4f8a-8f67-43ab35123684.png)

Yes, thanks! I think it may need a drop shadow on the main modal (perhaps even a border-radius), but it looks good!

TotalDarkness-NRF commented:
Changed it to look more like a window
![Screenshot_20210811_071949](https://user-images.githubusercontent.com/63280150/129020394-a6fb536e-02bb-4b07-9c03-eb128fa17d87.png)
![Screenshot_20210811_071906](https://user-images.githubusercontent.com/63280150/129020391-b961de54-f010-4c28-bc02-4ae778b4ecd4.png)

xou816 commented:
This looks a bit better already :)

Could you make it [modal](https://docs.gtk.org/gtk3/property.Window.modal.html)? I think it might fix some of the remaining visual issues

Also don't use HdyApplicationWindow for a "secondary" window -- HdyWindow should be enough :) and we won't have to have a headerbar too, which I think is unneeded here! Maybe for consistency we could have the same layout as the login dialog? (no "OK" would be needed either, with the close button being there)

jacksongoode commented:
> Changed it to look more like a window

Beautiful! would it also be possible to add a touch of padding around the info box so the text and top of the album art don't border the edges of the window (as well as the close button)?

xou816 commented:
Other general remarks, haven't properly dived into the code but:
- try not to mix UI models (such as AlbumDescription) with API models; the former should be allowed to evolve to fit our needs, while the latter are direct mappings of what the Spotify API exposes;
- the new AlbumInfo is not needed: just add the missing fields to the Album API model, that's what it's here for, and then map these new fields into AlbumDescription

xou816 commented:
Oh and I forgot but, the + XXms is really too much I think sorry :p 

Otherwise good job overall, I see you even handled translation strings I think!

TotalDarkness-NRF commented:
> This looks a bit better already :)
> 
> Could you make it [modal](https://docs.gtk.org/gtk3/property.Window.modal.html)? I think it might fix some of the remaining visual issues
> 
> Also don't use HdyApplicationWindow for a "secondary" window -- HdyWindow should be enough :) and we won't have to have a headerbar too, which I think is unneeded here! Maybe for consistency we could have the same layout as the login dialog? (no "OK" would be needed either, with the close button being there)

I beleive I had made it a modal already.
I can look at the way login dialog is made and replicate it for this for consistency

> Oh and I forgot but, the + XXms is really too much I think sorry :p

Ah sorry. I had added that for my own interest, forgot to have it removed (it is uneeded)

TotalDarkness-NRF commented:
Added the same layout as the login dialog and added the suggested padding

![image](https://user-images.githubusercontent.com/63280150/129103207-ce173300-253c-40b3-a856-9e540317c53d.png)

xou816 commented:
Oh you're right, my bad, modal is set!

Okay so a few more things related to the modal:
- there's an extra viewport that's not needed
- you can probably remove the scrolled window/viewport; or if you keep, make sure the height is propagated naturally and not hardcoded with height-request if possible
- if you keep the scrolling window, remove the shadow
- don't forget to allow wrapping labels (wrap=True, wrap-mode=word-char)
- good call to have this clamp!! make sure the whole thing can fit in 360px

Also, the info/help button could use another icon; what do you think of `preferences-system-details-symbolic`? And the button should be large enough to be a proper circle :) 

I toyed a bit with it, here's what I have in mind:
![Capture d’écran de 2021-08-12 19-36-17](https://user-images.githubusercontent.com/6277494/129243192-7b83f0ea-b534-458b-9a30-435806711d65.png)

Sorry, I know that's a lot of feedback, but it takes time to get these UI things just right :)

TotalDarkness-NRF commented:
Thanks for the feedback, I will do the best to follow it. The issue I have been having when not hardcoding the width and height is that the window size always becomes the size of the spot application. What should I do to solve this?

TotalDarkness-NRF commented:
New look with the suggested changes. I think it looks a lot better!
![image](https://user-images.githubusercontent.com/63280150/129259179-87b50928-a100-4ad0-8e42-83a69a52e92a.png)

xou816 commented:
Oh you're right my bad, but I think you can leave the height unset! (The width however should definitely be set to something <360 or should allow being resided to that size)

TotalDarkness-NRF commented:
The window width should be roughly the same each time, and the height will change
![image](https://user-images.githubusercontent.com/63280150/129284390-572e31f5-c0ff-4e10-90ed-c9147706d5b5.png)
![image](https://user-images.githubusercontent.com/63280150/129284513-875a5a28-4cf9-44a3-bef3-cb3331fa3d4c.png)

xou816 commented:
Neat!

Couple more things from testing:
- the top/bottom/start/end margins could be added to the first GtkBox, so the close button has some space around it; maybe this'll allow removing other margins further down 
![Capture d’écran du 2021-08-13 13-37-34](https://user-images.githubusercontent.com/6277494/129351742-f423c6a7-3526-4062-87cb-634777bb8417.png)

- the dialog should be resizable (this will allow mobile DEs to resize it to I believe), might have to tweak a few more things to allow it to shrink below 360px if needed
- I'm not too fond of the placement of the new button, especially on narrow screens (but thank you for updating the icon!); what would you think about either moving it below album+artist, or having it overlayed on top of the cover art? (in a corner for instance) maybe it should be flat as well, so that it appears as having a clearly different purpose from the "like" button (as in, not an action, just extra info)
![Capture d’écran du 2021-08-13 13-37-13](https://user-images.githubusercontent.com/6277494/129351760-04ee874e-e627-42ca-8a6c-019b2637af05.png)

TotalDarkness-NRF commented:
Yes I should try setting the margins to the gtkbox!
If I were to make the box resizable I may need to have the info in a scolled window. If there is something else that could be done please let me know! The width right now is almost always slightly bigger then the player controls, and the height grows.
The buttons may be moved to the corner of the album art because moving it under the album artist text would look more awkward.

xou816 commented:
> If I were to make the box resizable I may need to have the info in a scolled window.

Not necessarily, but you're right, better to be careful :)

> The width right now is almost always slightly bigger then the player controls

I tried making the window resizable but I wasn't able to shrink below a certain point :/ I'll take proper time to investigate but maybe one of the labels is not allowed to break?

> The buttons may be moved to the corner of the album art because moving it under the album artist text would look more awkward.

:+1: sounds good to me :) I'll think about it too, and will try a couple ideas... maybe centered at the bottom? Idk

TotalDarkness-NRF commented:
I think the issue would have to be the album artist label, as I specified it to be a certain length always to have it be a more consistant width each time. I will have to test some other ways of doing so with a resizable window

TotalDarkness-NRF commented:
I have added the overlay button on the album art. Let me know how it looks and what should be changed

Made the window resizable and scrollable and always start at 250 width.
![image](https://user-images.githubusercontent.com/63280150/129435761-3fa54c4d-2ec9-477c-8df4-dfb0ab5debd0.png)

xou816 commented:
Almost perfect, great work!! 

Couple suggested tweaks (to improve a bit legibility)
```diff
diff --git a/src/app/components/details/details.css b/src/app/components/details/details.css
index 9be249a..3300ae2 100644
--- a/src/app/components/details/details.css
+++ b/src/app/components/details/details.css
@@ -38,3 +38,7 @@ list.details__songs {
 list.details__songs row {
   border-radius: 4px;
 }
+
+.info_btn {
+  padding: 0;
+}
\ No newline at end of file
diff --git a/src/app/components/details/details.ui b/src/app/components/details/details.ui
index c0d7c1c..c625863 100644
--- a/src/app/components/details/details.ui
+++ b/src/app/components/details/details.ui
@@ -5,32 +5,34 @@
   <requires lib="libhandy" version="1.0"/>
   <!-- interface-css-provider-path details.css -->
   <object class="GtkButton" id="album_info">
-    <property name="width-request">35</property>
-    <property name="height-request">35</property>
+    <property name="width-request">32</property>
+    <property name="height-request">32</property>
     <property name="visible">True</property>
     <property name="can-focus">False</property>
     <property name="focus-on-click">False</property>
     <property name="receives-default">True</property>
-    <property name="halign">start</property>
-    <property name="valign">start</property>
-    <property name="relief">none</property>
+    <property name="halign">end</property>
+    <property name="valign">end</property>
+    <property name="margin">5</property>
+    <property name="relief">half</property>
     <child>
       <object class="GtkImage">
         <property name="visible">True</property>
         <property name="can-focus">False</property>
         <property name="icon-name">preferences-system-details-symbolic</property>
+        <property name="pixel-size">15</property>
       </object>
     </child>
     <style>
       <class name="circular"/>
+      <class name="osd"/>
+      <class name="info_btn"/>
     </style>
   </object>
```

I'll take a proper look at the code now thanks! :)

xou816 commented on src/app/models.rs:
```diff
@@ -86,6 +89,25 @@ impl AlbumDescription {
             .collect::<Vec<String>>()
             .join(", ")
     }
+    pub fn formatted_time(&self) -> String {
```
nice :) can probably be merged with the existing fn in utils

xou816 commented on src/app/state/screen_states.rs:
```diff
@@ -400,31 +400,4 @@ mod tests {
         let next = artist_state.next_page;
         assert_eq!(None, next.next_offset);
     }
-
```
hmm why? :p

TotalDarkness-NRF replied:
Must have deleted by accident when adding fieldsto albumdescription. woops sorry will re add

xou816 commented:
You'll also have to regenerate the potfiles: https://github.com/xou816/spot#regenerating-potfiles

TotalDarkness-NRF commented on src/app/components/details/details.rs:
```diff
@@ -54,6 +65,31 @@ impl Details {
+        });
+
+        widget.album_info.connect_clicked(move |_| {
+            // Stop from resizing when showing
```
This is an issue I have been having. When opening the info modal it will resize to the size of the spot application

TotalDarkness-NRF replied:
I think I figured out the issue! At some point I added a rule to spot in my window manager to remember last size and to ignore requested sizes. This was causing the window to always be the same size. Will remove now!

TotalDarkness-NRF commented on src/app/components/details/details.rs:
```diff
@@ -100,6 +136,57 @@ impl Details {
+                        .load_remote(&art[..], "jpg", 200, 200)
+                        .await;
+                    widget.info_art.set_from_pixbuf(pixbuf.as_ref());
+                    widget.set_loaded();
```
The loading of the image or the set loaded?

xou816 replied:
the `set_loaded` :)

TotalDarkness-NRF commented on src/app/components/details/details.rs:
```diff
@@ -100,6 +136,57 @@ impl Details {
+                .set_text(&format!("{}: {}", gettext("Label"), album.label));
+
+            self.widget.info_release.set_text(&format!(
+                "{}: {}",
```
Alright was not too sure thanks

TotalDarkness-NRF commented on src/app/components/details/details.ui:
```diff
@@ -4,6 +4,243 @@
+                                    <property name="can-focus">False</property>
+                                    <property name="halign">center</property>
+                                    <property name="valign">center</property>
+                                    <property name="label" translatable="yes">Album by Artist</property>
```
Should I remove translatable from this?

xou816 replied:
either remove or set it to no, both should work i think?

TotalDarkness-NRF commented on src/app/components/details/details.rs:
```diff
@@ -54,6 +65,31 @@ impl Details {
             }));
 
+        // Place the icon button above the album_art
+        widget.overlay.add_overlay(&widget.album_info);
```
Alright I could not figure it out earlier. Thanks!

TotalDarkness-NRF commented on src/app/models.rs:
```diff
@@ -75,6 +75,9 @@ pub struct AlbumDescription {
     pub artists: Vec<ArtistRef>,
     pub art: Option<String>,
     pub songs: Vec<SongDescription>,
+    pub label: String,
```
It has broken more then the artists page. I have to look at whats wrong. I think it may have to do with artist_album not including label and copyright info

TotalDarkness-NRF replied:
I think I fixed for search and artist page, now I discovered that this is a problem too with playlist as those dont list release dates. This is an issue with adding the fields label, release_date, and copyrights to Album

xou816 replied:
ok I see! if that is easier, we could have separate Album API model for endpoints that return a "light" version of it. that might be easier to handle than all the Options you'll have to deal with!

TotalDarkness-NRF replied:
Ye probably. Earlier on I had a struct that had all the information that AlbumDescription had + the extra info. Seeing that this info is only relevant for this screen maybe I could make it hold AlbumDetailedDescription that contains the details, and the extra info

TotalDarkness-NRF commented:
> You'll also have to regenerate the potfiles: https://github.com/xou816/spot#regenerating-potfiles

Alright. will do. Also I thought this only applied to when I would change the potfiles

TotalDarkness-NRF commented on src/app/components/details/details.ui:
```diff
@@ -4,6 +4,243 @@
+  <object class="HdyWindow" id="info_window">
+    <property name="can-focus">False</property>
+    <property name="modal">True</property>
+    <property name="window-position">center-always</property>
```
I have changed it to postion is none and gravity is North west. I am not sure if this is wanted though

xou816 replied:
i'll give it a try!

xou816 commented:
I'll have a proper look/review ASAP, thank you!!!

xou816 commented on src/api/cached_client.rs:
```diff
@@ -35,6 +35,8 @@ pub trait SpotifyApiClient {
 
     fn get_album(&self, id: &str) -> BoxFuture<SpotifyResult<AlbumDescription>>;
 
+    fn get_album_info(&self, id: &str) -> BoxFuture<SpotifyResult<AlbumDescriptionInfo>>;
```
Sorry, I should have been clearer; I stand by what I said earlier, there is no need to duplicate the endpoint here; there is no reason get_album and get_album_info would return different things

What I meant when I said that you could have multiple API models was that other models that depend on Album could depend on a different model, for instance (model names are just an example):
- RawSearchResults could have a Option<Page<LightAlbum>> because we don't get copyrights, label, etc
- the regular endpoint would return a FullAlbum
And then those two would be converted to our local model (AlbumDescription) or another one

xou816 commented on src/app/models.rs:
```diff
@@ -96,6 +100,35 @@ impl PartialEq for AlbumDescription {
 impl Eq for AlbumDescription {}
 
+#[derive(Clone, Debug)]
+pub struct AlbumDescriptionInfo {
```
good idea to compose :) 

naming suggestion: AlbumFullDescription ?

xou816 commented on src/app/models.rs:
```diff
@@ -96,6 +100,35 @@ impl PartialEq for AlbumDescription {
+}
+
+#[derive(Clone, Debug)]
+pub struct AlbumInfoRef {
```
a note on naming: those models I suffixed with Ref where meant to act like a "reference" to some resource; for instance ArtistRef would hold an identifier that would allow having a full Artist struct,  and a name to help display that reference

In that case, maybe AlbumReleaseDetails?

xou816 commented on src/app/models.rs:
```diff
@@ -96,6 +100,35 @@ impl PartialEq for AlbumDescription {
+pub struct AlbumInfoRef {
+    pub label: String,
+    pub release_date: String,
+    pub copyrights: Vec<CopyrightRef>,
```
(same kind of comment)

xou816 commented on src/api/cached_client.rs:
```diff
@@ -324,7 +324,17 @@ impl SpotifyApiClient for CachedSpotifyClient {
-            Ok(album)
+            let info = self
+                .cache_get_or_write(SpotCacheKey::Album(&id), None, |etag| {
+                    self.client.get_album_info(&id).etag(etag).send()
```
Almost there :)

See the issue here is that we end up calling two times the same album endpoint within this function -- it's alright because of the cache but it should not happen regardless (actually not fine, it writes two different models to the same cache entry but anyway)

Ask yourself, what does the v1/album endpoint actually return? a "full" album as defined by Spotify, so the union of your current AlbumInfo + the previous Album model

Just update the get_album fn from the client to return the full API model, and then you're good I think, this line is no longer needed :) nor is get_album_info needed

xou816 commented on src/app/models.rs:
```diff
@@ -96,6 +100,35 @@ impl PartialEq for AlbumDescription {
+#[derive(Clone, Debug)]
+pub struct AlbumFullDescription {
+    pub description: AlbumDescription,
+    pub info: AlbumReleaseDetails,
```
(nitpick: this field could be named release_details)

xou816 commented:
Thank you!! Looks like we're done :) I'll give it a quick final test to see if nothing broke after all the changes from review but it should be good to merge

ondras12345 commented:
Hi,
It looks like there's something wrong with the positioning of the dialog window.
When I open some album's info and try moving the main Spot window by dragging it's titlebar, the dialog window jumps to the right edge of my rightmost monitor.
The interesting thing is that when I hold down `Super` while moving the main window, everything seems to work just fine.
![Spot-album-info-move](https://user-images.githubusercontent.com/44524735/129889329-35f3335b-93c2-43da-8820-998962d4c1c2.gif)

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 20.04.2 LTS
Release:	20.04
Codename:	focal
```
GNOME 3.36.8

Do you want me to create an issue for this?

xou816 commented:
Hi! Yes, an issue will be easier to track, thank you! :)
